luarules/gadgets/unit_area_timed_damage.lua

```diff
@@ -16,225 +16,521 @@ if not gadgetHandler:IsSyncedCode() then
-		}
+	local duration = weaponParams.area_duration
+	local radius = weaponParams.area_radius
+	local elevation = max(0, spGetGroundHeight(x, z))
 
-		Spring.SpawnCEG(
-			params.atd_stable_ceg,
+	-- Create an area on surface -- immediately.
+	if y <= elevation + radius then
+		local group = timedAreas[frame]
+		group[#group+1] = {
+			duration + time,
+			weaponParams,
+			ownerID,
 			x, elevation, z,
-			0, 0, 0,
-			params.atd_wdef.damageAreaOfEffect, -- Just used for scaling?
-			params.atd_wdef.damages[0]          -- Just used for scaling?
-		)
-	end
-end
+		}
+
+		-- Most timed areas are represented by a long-duration CEG.
+		-- The others use an explosiongenerator on the area weapon.
+		if weaponParams.area_ongoingceg then
+			spSpawnCEG(weaponParams.area_ongoingceg, x, elevation, z)
+		end
+
+	-- Create an area on surface -- eventually.
+	elseif shortDuration < duration then
+		local timeToLand = sqrt((y - elevation - radius) * 2 / gravity)
+		-- This check considers two different limits on the fall time:
+		-- (1) The max latency players associate a cause (impact) to an effect (area on ground); ~1 sec.
+		-- (2) The max time that the falling area effect might remain cohesive; just a heuristic formula.
+		if timeToLand < 1 and timeToLand <= 0.25 + (duration - shortDuration) * 0.1 then
+			local frameStart = spGetGameFrame() + ceil(timeToLand / gameSpeed) -- at least +1 frame
+			if not delayQueue[frameStart] then
+				delayQueue[frameStart] = { x, elevation, z, weaponParams, ownerID }
+			else
+				local queue = delayQueue[frameStart]
+				local sizeq = #queue
+				queue[sizeq + 1] = x
+				queue[sizeq + 2] = elevation
+				queue[sizeq + 3] = z
+				queue[sizeq + 4] = weaponParams
+				queue[sizeq + 5] = ownerID
+			end
 
-local function StopTimedArea(group, index)
-	if index == #group then
-		group[#group] = nil
-	else
-		group[index] = empty
-		freed[#freed + 1] = index
+		-- Spawn a single explosion in mid-air with no recurring area.
+		else
+			local explosion = explosionCaches[weaponParams]
+			explosion.owner = ownerID ~= -1 and ownerID or nil
+			spSpawnExplosion(x, y, z, 0, 0, 0, explosion)
+		end
 	end
 end
 
-local function UpdateTimedAreas(frame)
-	local frameAreas = areas[frame % 30]
-	local frameRate = Spring.GetGameSpeed()
-	local frameTime = Spring.GetGameSeconds() - 0.5 / frameRate
-	-- local damageRate = 30 / frameRate -- todo
+local function updateTimedAreas()
+	local group = timedAreas[frame]
+	local sizeg = #group
 
-	local explosion = {
-		weaponDef          = 0, -- todo: Great. Now I gotta make a weapon for all of these.
-		edgeEffectiveness  = 1,
-		explosionSpeed     = 10000,
-		damageGround       = false,
-	}
+	local index = 1
+	while index <= sizeg do
+		local timedArea = group[index]
+		if time <= timedArea[1] then
+			local explosion = explosionCaches[timedArea[2]]
+			explosion.owner = timedArea[3] ~= -1 and timedArea[3] or nil 
+			spSpawnExplosion(timedArea[4], timedArea[5], timedArea[6], 0, 0, 0, explosion)
```

Comment:
Numeric table indexes are cryptic and make reading code difficult, better to use named indexes.

Reply:
I was torn over the extra cost to use the hashed part of tables, but consider this poor practice now. Will fix.